Source/WebCore/Modules/permissions/Permissions.cpp

```diff
@@ -77,66 +82,71 @@ static bool isAllowedByFeaturePolicy(const Document& document, PermissionName na
+            return;
+        }
     }
 
     JSC::VM& vm = context->globalObject()->vm();
     auto scope = DECLARE_THROW_SCOPE(vm);
-    auto parameterDescriptor = convert<IDLDictionary<PermissionDescriptor>>(*context->globalObject(), permissionDescriptorValue.get());
+    auto permissionDescriptor = convert<IDLDictionary<PermissionDescriptor>>(*context->globalObject(), permissionDescriptorValue.get());
     if (UNLIKELY(scope.exception())) {
         promise.reject(Exception { ExistingExceptionError });
         return;
     }
 
-    if (is<Document>(context) && !isAllowedByFeaturePolicy(downcast<Document>(*context), parameterDescriptor.name)) {
-        context->postTask([parameterDescriptor, promise = WTFMove(promise)](auto& context) mutable {
-            promise.resolve(PermissionStatus::create(context, PermissionState::Denied, parameterDescriptor));
-        });
+    if (document && !isAllowedByFeaturePolicy(*document, permissionDescriptor.name)) {
+        promise.resolve(PermissionStatus::create(*context, PermissionState::Denied, permissionDescriptor));
         return;
     }
 
     auto* origin = context->securityOrigin();
     auto originData = origin ? origin->data() : SecurityOriginData { };
 
-    PermissionController::shared().query(ClientOrigin { context->topOrigin().data(), originData }, PermissionDescriptor { parameterDescriptor }, *document->page(), [this, protectedThis = Ref { *this }, parameterDescriptor, promise = WTFMove(promise)](auto permissionState) mutable {
-        auto context = m_navigator ? m_navigator->scriptExecutionContext() : nullptr;
-        if (!context || !context->globalObject())
-            return;
+    if (auto* workerGlobalScope = dynamicDowncast<WorkerGlobalScope>(*context)) {
+        auto completionHandler = [originData = WTFMove(originData).isolatedCopy(), permissionDescriptor, contextIdentifier = workerGlobalScope->identifier(), promise = WTFMove(promise)] (auto& context) mutable { 
+            ASSERT(isMainThread());
 
-        context->postTask([parameterDescriptor, promise = WTFMove(promise),  permissionState = WTFMove(permissionState)](auto& context) mutable {
-            if (!permissionState) {
-                promise.reject(Exception { NotSupportedError });
+            auto& document = downcast<Document>(context);
+            if (!document.page()) {
+                promise.reject(Exception { InvalidStateError, "The page does not exist"_s });
                 return;
             }
-            promise.resolve(PermissionStatus::create(context, *permissionState, parameterDescriptor));
-        });
+
+            PermissionController::shared().query(ClientOrigin { context.topOrigin().data(), WTFMove(originData) }, PermissionDescriptor { permissionDescriptor }, *document.page(), [context = Ref { context }, contextIdentifier, permissionDescriptor, promise = WTFMove(promise)](auto permissionState) mutable {
+                ASSERT(context.get().globalObject());
```

Comment:
Why? How is this a guarantee? Note that you are making changes I didn't request between iterations and it is making it hard to review.

Reply:
You're right--this isn't a guarantee. Fixed!